google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkProgressListener.java

```diff
@@ -0,0 +1,187 @@
+    }
+  }
+
+  @Override
+  public void update(String message) {
+    setProgressText(message);
+  }
+
+  @Override
+  public void done() {
+    if (!progressIndicator.isFinished(task)) {
+      progressIndicator.finish(task);
+      progressIndicator.dispose();
+    }
+  }
+
+  @Override
+  public ProgressListener newChild(long allocation) {
+    return new ChildProgressListener(allocation);
+  }
+
+  private void setProgressText(String message) {
+    progressIndicator.setText(GctBundle.message("managedsdk.progress.message", message));
+  }
+
+  @VisibleForTesting
+  BackgroundableProcessIndicator initProgressIndicator() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // task info for progress indicator, required to init/finish properly.
+              task =
+                  new Backgroundable(
+                      null /* not project specific task */,
+                      GctBundle.message("managedsdk.notifications.title"),
+                      true /* cancellable */,
+                      PerformInBackgroundOption.ALWAYS_BACKGROUND) {
+                    @Override
+                    public void run(@NotNull ProgressIndicator indicator) {
+                      // will not be called in manual progress mode.
+                    }
+                  };
+              progressIndicator = new BackgroundableProcessIndicator(task);
+              Disposer.register(ApplicationManager.getApplication(), progressIndicator);
+              progressIndicator.addStateDelegate(
+                  new AbstractProgressIndicatorExBase() {
+                    @Override
+                    public void cancel() {
+                      managedCloudSdkService.cancelInstallOrUpdate();
+                      done();
+                    }
+                  });
+            });
+
+    return progressIndicator;
+  }
+
+  /**
+   * Progress listener for child activities, updates the same progress indicator. Updates progress
+   * indicator with proportioned chunk of global progress.
+   */
+  private final class ChildProgressListener implements ProgressListener {
+
+    private double childWork;
+    /* 0 < childTotalWorkAllocation <= totalWork */
+    private final double childTotalWorkAllocation;
+
+    private ChildProgressListener(double childTotalWorkAllocation) {
+      this.childTotalWorkAllocation = childTotalWorkAllocation;
+    }
+
+    @Override
+    public void start(String message, long childWork) {
+      this.childWork = childWork;
+
+      setProgressText(message);
+
+      if (childWork == UNKNOWN) {
+        // pass unknown state to parent listener.
+        ManagedCloudSdkProgressListener.this.update(UNKNOWN);
+      }
+    }
+
+    @Override
+    public void update(long childWorkDone) {
+      // both could be UNKNOWN meaning undetermined total progress for the whole child task.
+      if (childWorkDone != UNKNOWN && childWork != UNKNOWN) {
```

Comment:
its a little tough to understand the difference between `childWorkDone` and `childWork`. maybe this can be solved with naming

Reply:
also how they differ from `childTotalWorkAllocation `

Reply:
Lol, I knew you would ask me about these. I renamed them myself at least three times. I hated the concept of "child progress showing chunk of global progress" instead of just progress bar showing 0 to 100% done, but alas, Eclipse team voted for this. 

So global progress amount is `totalWork`. When child progress is created, it's passed `childTotalWorkAllocation` - chunk of `totalWork` done by this child (i.e. totalWork = 100, childTotalWorkAllocation = 50 for half of it.). Then child `start` is called with `childWork` - this is arbitrary number of child work, completely unrelated to `totalWork` and `childTotalWorkAllocation`. It's only related to `update(childWorkDone)`, which updates work until all of `childWork` is done.

If you could help me with better names for child numbers for work unrelated to global total work, that'd be great, I am kind of confused with better names here. Maybe `childWorkDone` -> `childWorkChunkCompleted` or similar?

Reply:
oh man. im really getting tested this morning :). I'll try to take this in and make a suggestion if I have one.

Reply:
Yea this is very bizarre. Some thoughts:
 
`totalWork` - all the work ever to be done. name seems fine.
`childTotalWorkAllocation` - fraction of the total work completed by this "child"? maybe `totalWorkCompleted`?
`childWork` - some arbitrary number reflecting the total amount of work of this child? maybe `totalChildWork`
`childWorkDone` - the fraction of the total child work completed? maybe `totalChildWorkCompleted`

assuming my understanding is correct. now you have:

`totalWork` and `totalWorkCompleted`
`totalChildWork` and `totalChildWorkCompleted`

making these two analogous concepts have consistent naming.

Reply:
Sounds good, I like the analogy between them. The only thing is `totalWorkCompleted ` for total work part allocated to a child, it's not yet completed when assigned and it sounds like it's what is already done now globally... how about `totalWorkAssigned` maybe?

Reply:
ah I see why you went with `allocation` now. `totalWorkAssigned` or `totalWorkAllocated` sounds good to me

Reply:
@ivanporty can't you use the `appengine-plugin-core`'s [`ChildProgressListener`](https://github.com/GoogleCloudPlatform/appengine-plugins-core/blob/master/src/main/java/com/google/cloud/tools/managedcloudsdk/ChildProgressListener.java)?

Reply:
@briandealwis I'll take a look, thanks

Reply:
@briandealwis thanks for the suggestion, it looks like it does exactly what we did here with minor deviations, so we get rid of extra code.

@etanshaul PTAL, I think now it's much easier when child stuff is hidden )